### PR TITLE
feature/MS/postulants: Postulant entity created. Methods, Routes and json added.

### PR DESCRIPTION
**Pull request by clementinob**
PR to be added:

- Routes to postulants --> inside `./index.js`.

- Controllers: Methods to get all Postulants, add one new Postulant, delete or edit an existing Postulant, search by id or last name -->  inside `./controllers/postulants.js`.

- Data of Postulants --> inside `./data/postulants.json`.

### CODE REVIEW

File: data/postulants.json [hunk 1]
```diff
@@ -0,0 +1 @@
...gender","date_of_birth":"09/21/1977","email":"kfust1d@weather.com","phone":"(877) 4295681","address":"91053 Atwood Avenue","city":"Sabaneta","state":null,"country":"Dominican Republic","elementary_school":{"name":"Universidad Abierta Para Adultos","graduate_year":"11/23/2007"},"high_school":{"name":"Universidad Tecnológica del Sur","graduate_year":"03/23/2009"},"junior_college":{"name":"Universidad Central Dominicana de Estudio Profesionales","degree":"Software Test Engineer I","graduate_year":"07/23/1984"},"university":{"name":"Universidad Católica de Santo Domingo","degree":"Junior Executive","graduate_year":"02/13/1998"},"open_to_work":true,"work_experience":{"title":"Product Engineer","start":"04/05/2002","end":"03/25/1992","company":"Ozu","desription":"ligula suspendisse ornare consequat lectus in est risus auctor sed tristique in tempus sit amet sem fusce consequat nulla nisl"},"professional_training":{"description":"Public Policy","year":"08/28/1977"},"languages":["Telugu","Macedonian","Bengali"],"hobbies":"pellentesque ultrices mattis odio donec","family_members":{"name":"Koenraad","bond":"non"},"availability":{"from":13,"to":16,"day":[null,null,null,null,null,null,null]}}]
```
Comment (aquilesmariani) [inline data/postulants.json:1]: It would be nice if we have the mock formatted and not as a single line file

File: controllers/postulants.js [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  }
+  fs.writeFile('./data/postulants.json',JSON.stringify(Postulants),err => {
+    if (err) { res.status(500).json({ msg: `Error while saving the Postulants with the updated Postulant` });
+    } else {
+    res.json({ msg: 'Postulant updated!', postulantToUpdate});
+    }})
+};
+
+const deletePostulant = (req, res) => {
+  const found = Postulants.some(postulant => postulant.id === parseInt(req.params.id));
```
Comment (aquilesmariani) [inline controllers/postulants.js:184]: Could you replace using `some` and `findIndex` by just using `findIndex` and checking that it exists just if `indexToBeDeleted >= 0`?

File: controllers/postulants.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+const updatePostulant = (req, res) => {
+  let postulantPosition = -1;
+  let postulantToUpdate = Postulants.find((postulant, index) => {
+    if (postulant.id === parseInt(req.params.id)) {
+      postulantPosition = index;
+      return postulant;
+    }
+  });
+  if (postulantToUpdate) {
+    postulantToUpdate = {
```
Comment (aquilesmariani) [inline controllers/postulants.js:109]: Nice! This was a huge object to work on! Appreciated 💯

File: controllers/postulants.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+  const found = Postulants.some(postulant => postulant.id === parseInt(req.params.id));
+    if (found) {
+        res.json(Postulants.filter(postulant => postulant.id === parseInt(req.params.id)))
+    } else {
+        res.status(400).json({ msg: 'Postulant not found' })
+    }
+}
+
+const getPostulantsByLastName = (req, res) => {
+  const found = Postulants.some(postulant => postulant.last_name === req.params.last_name);
```
Comment (aquilesmariani) [inline controllers/postulants.js:92]: Same thing I mentioned about this. Try using just one method to iterate through the array (in this case using filter) so you don't go over it more than just once.

File: controllers/postulants.js [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+        ],
+    }
+  }
+  
+  Postulants.push(newPostulant);
+  fs.writeFile('./data/postulants.json', JSON.stringify(Postulants), (err) => {
+    if (err) {
+      res.status(400).json({msg: "Error. Check the data sended and try again."});
+    }else{
+      res.status(200).json({msg: `The Postulant with the ID: ${newPostulant.id} has been created.`});
```
Comment (aquilesmariani) [inline controllers/postulants.js:77]: This works fine. But for creating a new resource, I think it's better that we return a 201 status code.

File: controllers/postulants.js [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+          req.query.availability_day4, req.query.availability_day5, req.query.availability_day6, 
+          req.query.availability_day7
+        ],
+    }
+  }
+  
+  Postulants.push(newPostulant);
+  fs.writeFile('./data/postulants.json', JSON.stringify(Postulants), (err) => {
+    if (err) {
+      res.status(400).json({msg: "Error. Check the data sended and try again."});
```
Comment (aquilesmariani) [inline controllers/postulants.js:75]: Also, there's no need to return a message as a json format inside of an object. We could just a string instead:

```suggestion
        res.status(404).send("Error. Check the data sent and try again.");
```

File: controllers/postulants.js [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+      }
+  })
+}
+
+const getPostulantsById = (req, res) => {
+  const found = Postulants.some(postulant => postulant.id === parseInt(req.params.id));
+    if (found) {
+        res.json(Postulants.filter(postulant => postulant.id === parseInt(req.params.id)))
+    } else {
+        res.status(400).json({ msg: 'Postulant not found' })
```
Comment (aquilesmariani) [inline controllers/postulants.js:87]: The status code for when the requested resource hasn't been found is `404`.

**Review by aquilesmariani — COMMENTED** [review]
Way to go, Bruno! Just some adjustments and we'll be ready to merge

File: data/postulants.json (hunk 1 above)
Comment (clementinob) [reply data/postulants.json:1]: Done! Sorry about that. The FileSystem make a mess with that poor json while testing.

File: controllers/postulants.js (hunk 2 above)
Comment (clementinob) [reply controllers/postulants.js:184]: Perfect! Thanks for pointing me that. Now all the methods only iterate once the file.

File: controllers/postulants.js (hunk 3 above)
Comment (clementinob) [reply controllers/postulants.js:109]: Thanks Aqui! It was a little bit complicated in the beggining, but nothing impossible.

File: controllers/postulants.js (hunk 4 above)
Comment (clementinob) [reply controllers/postulants.js:92]: Done. :)

File: controllers/postulants.js (hunk 5 above)
Comment (clementinob) [reply controllers/postulants.js:77]: Great! Done! ✔️

File: controllers/postulants.js (hunk 6 above)
Comment (clementinob) [reply controllers/postulants.js:75]: Noted! Now all the messages are a string. 👍

File: controllers/postulants.js (hunk 7 above)
Comment (clementinob) [reply controllers/postulants.js:87]: Done! Now all search validations returns that status code.

**Review by aquilesmariani — APPROVED** [review]
LGTM 🚀

**Pull request merged**